include/mapbox/platform.hpp

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+// Determine compiler
+#define MB_COMPILER 0
+#define MB_COMPILER_GNU 1
+#define MB_COMPILER_CLANG 2
+#define MB_COMPILER_MSVC 3
+
+#if defined(__clang__)
+#    undef MB_COMPILER
+#    define MB_COMPILER MB_COMPILER_CLANG
+#elif defined(__GNUC__) || defined(__GNUG__) // after clang, possible it also has the same defines (need to research)
+#    undef MB_COMPILER
+#    define MB_COMPILER MB_COMPILER_GNU
+#elif defined(_MSC_VER)
+#    undef MB_COMPILER
+#    define MB_COMPILER MB_COMPILER_MSVC
+#else
+#    error "Unsupported compiler"
+#endif
+
+// Determine platform
+#define MB_PLATFORM 0
+#define MB_PLATFORM_WIN32 1
+#define MB_PLATFORM_LINUX 2
+#define MB_PLATFORM_MAC 3
+#define MB_PLATFORM_IOS_EMBEDDED 4
+#define MB_PLATFORM_IOS_SIMULATOR 5
+#define MB_PLATFORM_ANDROID 6
+
+#if (defined(__WIN32__) || defined(_WIN32)) && !defined(__ANDROID__)
+#    undef MB_PLATFORM
+#    define MB_PLATFORM MB_PLATFORM_WIN32
+#elif defined(__APPLE_CC__)
+#    include "TargetConditionals.h"
+#    undef MB_PLATFORM
+#    if TARGET_OS_IOS && TARGET_OS_EMBEDDED
+#        define MB_PLATFORM MB_PLATFORM_IOS_EMBEDDED
+#    elif TARGET_OS_IOS && TARGET_OS_SIMULATOR
+#        define MB_PLATFORM MB_PLATFORM_IOS_SIMULATOR
+#    else
+#        define MB_PLATFORM MB_PLATFORM_MAC
+#    endif
+#elif defined(__ANDROID__)
+#    undef MB_PLATFORM
+#    define MB_PLATFORM MB_PLATFORM_ANDROID
+#else // let all other platform right now is a linux platforms
+#    undef MB_PLATFORM
+#    define MB_PLATFORM MB_PLATFORM_LINUX
+#endif
+
+#define MB_PLATFORM_IS_WIN32 (MB_PLATFORM == MB_PLATFORM_WIN32)
+#define MB_PLATFORM_IS_LINUX (MB_PLATFORM == MB_PLATFORM_LINUX)
```

Comment:
I think we need to make a distinction between Linux Desktop and Linux Embedded here.

Reply:
I don't know how to do that.

Reply:
Maybe we could check whether we are cross-compiling or not, or look for specific platform preprocessor defines like `__amd64` or `__arm__`.

Reply:
Yep, we can figure out that and add in future